pallets/price-oracle/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+
+pub type AssetPairId = Vec<u8>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Weight information for the extrinsics.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Calculation error occurred while calculating average price
+		PriceComputationError,
+
+		/// An unexpected overflow occurred
+		UpdateDataOverflow,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Pool was registered. [AssetPair]
+		PoolRegistered(AssetPair),
+	}
+
+	/// The number of assets registered and handled by this pallet.
+	#[pallet::storage]
+	#[pallet::getter(fn num_of_assets)]
+	pub type NumOfTrackedAssets<T: Config> = StorageValue<_, u32, ValueQuery>;
+
+	/// Processed or partially processed data generated by trades.
+	/// Data generated by trades are processed sequentially.
+	/// Each new entry is combined with the previous value to produce new intermediate value.
+	/// The last entry creates the resulting average price and volume.
+	#[pallet::storage]
+	#[pallet::getter(fn price_accumulator)]
+	pub type PriceDataAccumulator<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, PriceEntry, ValueQuery>;
+
+	/// The last ten average values corresponding to the last ten blocks.
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_ten)]
+	pub type PriceDataTen<T: Config> = StorageValue<_, Vec<(AssetPairId, BucketQueue)>, ValueQuery>;
+
+	/// The last ten average values corresponding to the last hundred blocks.
+	/// Each average value corresponds to an interval of length ten blocks.
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_hundred)]
+	pub type PriceDataHundred<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, BucketQueue, ValueQuery>;
+
+	/// The last ten average values corresponding to the last thousand blocks.
+	/// Each average value corresponds to an interval of length hundred blocks.
+	#[pallet::storage]
+	#[pallet::getter(fn price_data_thousand)]
+	pub type PriceDataThousand<T: Config> = StorageMap<_, Blake2_128Concat, AssetPairId, BucketQueue, ValueQuery>;
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(_n: T::BlockNumber) -> Weight {
+			// update average values in the storage
+			Self::update_data();
+			// clear the price buffer
+			PriceDataAccumulator::<T>::remove_all(None);
+
+			T::WeightInfo::on_initialize_one_entry_multiple_tokens(Self::num_of_assets())
+		}
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+}
+
+impl<T: Config> Pallet<T> {
+	pub fn on_create_pool(asset_pair: AssetPair) {
+		let data = PriceDataTen::<T>::get();
+		if !data.iter().any(|bucket_tuple| bucket_tuple.0 == asset_pair.name()) {
```

Comment:
Yes it is necessary. We need to check the existence of an asset pair and increase the asset_count if the asset pair exists because it is used in weigh calculation here https://github.com/galacticcouncil/Basilisk-node/blob/534465213edee32e2a9ad24466bd1ebf1685b95a/pallets/price-oracle/src/weights.rs#L82

Reply:
yes, i know it is necessary now. I really meant if this could not be  done in a way where there is no need to iterate over it just to find the pair

Reply:
I'm not aware of a better solution. We need to use an iteration because we use a vector. The reason behind this is that we need to iterate over all values at the end of each block, and because it is advised to use vectors over maps (faster iterations). https://substrate.dev/docs/en/knowledgebase/runtime/storage#iterating-over-storage-maps

Reply:
why not to just use a map where this check could be O(1) ?